enhancements/installer/manifests-from-stdin.md

```diff
@@ -0,0 +1,114 @@
+reviewers:
+  - "@fzdarsky"
+approvers:
+  - "@crawford"
+creation-date: 2020-12-24
+last-updated: 2020-12-24
+status: provisional
+---
+
+# Manifests from STDIN
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In many occasions, users need to customize the install-config target and/or the manifests generated by the OpenShift Installer in a reproducible manner. The most popular tool to allow this type of customization is “Kustomize”, and it is already well-known by the Kubernetes ecosystem. The result of any “Kustomization” is sent to standard output as a stream of manifests.
+
+This OpenShift enhancement proposes to allow the installer to read a stream of manifests from STDIN on the main targets that consume manifests.
+
+
+## Motivation
+
+Infrastructure-as-Code (e.g. GitOps) is an operational best practice for provisioning and managing large numbers of clusters and services. A typical IaC workflow involves pulling Kubernetes resource manifest templates of a specific version from a `git` repo, rendering those templates for a given cluster instance, and then running `openshift-install` (resp. `oc apply`) on the rendered manifests to install (resp. update) a cluster.
+
+Pulling and rendering manifest templates is conveniently done via `kustomize` - which is particularly suited for an IaC workflow - but other tools like `helm` and `jsonnet` have their followers, too. These tools have in common that they render templates into a stream of manifests sent to `stdout`.
+
+As `openshift-install` currently only supports reading manifest files from the file system and requires files to follow a specific naming and directory structure convention, IaC workflows become cumbersome and often involve users writing their own Bash scripts or Ansible playbooks to automate the workflows. By giving the installer the ability to read from `stdin`, we will enable a simplified workflow pattern via pipelines while still being capable of scaling with the complexity of users’ needs.
+
+### Goals
+
+Enable the Openshift Installer to support pipelined workflows by reading manifests from STDIN as an alternative to reading from the filesystem.
+
+### Non-Goals
+
+- Change the existing workflow to deploy an OpenShift cluster.
+- Couple the Openshift Installer with Kustomize or other manifest tooling
+
+
+## Proposal
+
+Modify the openshift-install binary to accept input from STDIN for all targets which consume manifests.  For example:
+
+- Create manifests from customized install-config.yaml: `kustomize build github.com/myorg/myBlueprint/install-config-kustomization | openshift-install create manifests -f -`
```

Comment:
Is the intent to make it easier to pass *extra* manifests to the installer, or to make it easier to modify the manifests created by the installer?

Reply:
More the later - in the example, the yaml generated by the `kustomize build` would adhere to the current install-config scheme.  A more mundane/contrived example would be `openshift-install create install-config && cat install-config.yaml | openshift-install create manifests -f -`

The idea is that we'd be supporting a simple and powerful pipeline model which could for instance take install-configs created via kustomize.

Reply:
I guess the source of my confusion is that I didn't really consider the `install-config.yaml` file as "manifest" in the same way that the files created by `openshift-install create manifests`. Would you expect to be able to pass the output of `create manifests` through `kustomize`?

Reply:
The idea would be to have a generic `install-config.yaml` describing at a high level view, how the cluster looks like (number of masters and workers, flavor of VMs, etc.) and use kustomize to overwrite specific information from the site (cluster name, domain, ipmi addresses, etc.). Using the base `install-config.yaml` + kustomize patch, the result shall be use to generate the installer manifests.

This approach will give organizations the ability to use a common recipe or `install-config.yaml` and customize only certain parameters. This would be very useful in edge computing usecases at scale.

Reply:
I agree I was thrown off by 'manifests' I assumed it was limited in scope to the output of `openshift-install create manifests` perhaps we should be more generic and speak of "Installer inputs" or "Installer State"?

Reply:
Based on the examples of how this would be used, I like "installer inputs" because it differentiates from what the installer reads and what it produces as part of its work.

@oglok describes how install-config.yaml would be read. What about other inputs? Do we want it to be able to read extra content, like machine configs, from stdin as well?

Reply:
I'd assumed that https://github.com/openshift/enhancements/pull/575/files#diff-c776d0a2f45ebd59d0f77c302549387a07069285f130da08a3a66aa24553792cR56-R57 implied we were mutating the output of `openshift-install create manifests` and `openshift-install create ignition-configs` respectively in the full `install-config` `manifests` `ignition-configs` `cluster` create workflow. Maybe I've misunderstood.

Reply:
I can certainly replace "manifests" by "inputs". The workflow would require to apply "kustomizations" on the entire workflow. 

- Kustomize overlay (cluster name, number of nodes...) + install-config.yaml | `openshift-install create manifests`
- Kustomize overlay (adding MachineConfigs, patching IngressReplicas...) + openshift installer manifests | `openshift-install create (ignition-configs | cluster )`

If the user wants to deploy following the UPI approach, he/she would create the ignition files to inject in RHCOS. If the user wants to deploy using any IPI available option,  he/she would have to just create the cluster.

Reply:
I have limited experience with kustomize, but the only way I've ever seen it used requires the patch files and base files to be in a common directory hierarchy (for example, https://github.com/metal3-io/baremetal-operator/tree/master/config). I've had it report errors when there was not a common parent directory. So, IIUC, modifying the files produced by the installer would involve a separate step to set up the right directory hierarchy (by moving or copying files) before running kustomize.

Since we can already use a multi-step approach with no changes to the installer, I think that means we the proposed approach would only apply to things that are pure inputs to the installer, and we would be assuming that the changes made by kustomize would propagate from those inputs to the manifests the installer generates. A more detailed example of a use case, describing how kustomize would change something more complex than say the cluster name, would help me understand if that assumption is correct.

Reply:
From kustomize v3.7.0 it's possible to create reusable "kustomizations" called `Components`.

Let's say that according to this enhacenment proposal, the output of `openshift-install create manifests` generates in the chosen directory the following structure:

- openshift/
- manifests/
- kustomization.yaml (this file will list all manifests as resources)

and we want to apply some remote overlay on top of the manifests generated by the installer. We would need to edit the generated kustomization.yaml with the following `components` section:

```
resources:                                                                                                                                                                                                      
- manifests/openshift-config-secret-pull-secret.yaml                                                                                                                                                            
- manifests/machine-config-server-tls-secret.yaml                                                                                                                                                               
- manifests/kube-system-configmap-root-ca.yaml                                                                                                                                                                  
- manifests/kube-cloud-config.yaml                                                                                                                                                                              
- manifests/etcd-signer-secret.yaml                                                                                                                                                                             
- manifests/etcd-serving-ca-configmap.yaml                                                                                                                                                                      
- manifests/etcd-service.yaml                                                                                                                                                                                   
- manifests/etcd-namespace.yaml                                                                                                                                                                                 
- manifests/etcd-metric-signer-secret.yaml                                                                                                                                                                      
- manifests/etcd-metric-serving-ca-configmap.yaml                                                                                                                                                               
- manifests/etcd-metric-client-secret.yaml                                                                                                                                                                      
- manifests/etcd-host-service.yaml                                                                                                                                                                              
- manifests/etcd-host-service-endpoints.yaml                                                                                                                                                                    
- manifests/etcd-client-secret.yaml                                                                                                                                                                             
- manifests/etcd-ca-bundle-configmap.yaml                                                                                                                                                                       
- manifests/cvo-overrides.yaml                                                                                                                                                                                  
- manifests/cluster-scheduler-02-config.yml                                                                                                                                                                     
- manifests/cluster-proxy-01-config.yaml                                                                                                                                                                        
- manifests/cluster-network-02-config.yml                                                                                                                                                                       
- manifests/cluster-network-01-crd.yml                                                                                                                                                                          
- manifests/cluster-ingress-02-config.yml                                                                                                                                                                       
- manifests/cluster-infrastructure-02-config.yml                                                                                                                                                                
- manifests/cluster-dns-02-config.yml                                                                                                                                                                           
- manifests/cluster-config.yaml                                                                                                                                                                                 
- manifests/cloud-provider-config.yaml                                                                                                                                                                          
- manifests/cloud-controller-uid-config.yml                                                                                                                                                                     
- manifests/04-openshift-machine-config-operator.yaml                                                                                                                                                           
- openshift/99_openshift-machineconfig_99-master-ssh.yaml                                                                                                                                                       
- openshift/99_openshift-cluster-api_master-user-data-secret.yaml                                                                                                                                               
- openshift/99_openshift-cluster-api_master-machines-2.yaml                                                                                                                                                     
- openshift/99_openshift-cluster-api_master-machines-1.yaml                                                                                                                                                     
- openshift/99_openshift-cluster-api_master-machines-0.yaml                                                                                                                                                     
- openshift/99_openshift-machineconfig_99-worker-ssh.yaml                                                                                                                                                       
- openshift/99_openshift-cluster-api_worker-user-data-secret.yaml                                                                                                                                               
- openshift/99_openshift-cluster-api_worker-machineset-2.yaml                                                                                                                                                   
- openshift/99_openshift-cluster-api_worker-machineset-1.yaml                                                                                                                                                   
- openshift/99_openshift-cluster-api_worker-machineset-0.yaml                                                                                                                                                   
- openshift/openshift-install-manifests.yaml                                                                                                                                                                    
- openshift/99_role-cloud-creds-secret-reader.yaml                                                                                                                                                              
- openshift/99_kubeadmin-password-secret.yaml                                                                                                                                                                   
- openshift/99_cloud-creds-secret.yaml                                                                                                                                                                          
                                                                                                                                                                                                                
components:                                                                                                                                                                                                     
- git::https://github.com/oglok/blueprint-test.git//overlay-manifests/ 
```

The manifests located in that remote location will look like:

- etcd-service-port.yaml
- etcd-host-service-port.yaml
- kustomization.yaml

_**etcd-service-port.yaml**_
apiVersion: v1                                                                                                                                                                                                  
kind: Service                                                                                                                                                                                                   
metadata:                                                                                                                                                                                                       
  name: etcd                                                                                                                                                                                                    
  namespace: openshift-etcd                                                                                                                                                                                     
  labels:                                                                                                                                                                                                       
    k8s-app: etcd                                                                                                                                                                                               
spec:                                                                                                                                                                                                           
  ports:                                                                                                                                                                                                        
    - $patch: replace                                                                                                                                                                                           
    - name: etcd                                                                                                                                                                                                
      port: 2378                                                                                                                                                                                                
      protocol: TCP      

_**etcd-host-service-port.yaml**_
apiVersion: v1                                                                                                                                                                                                  
kind: Service                                                                                                                                                                                                   
metadata:                                                                                                                                                                                                       
  name: host-etcd                                                                                                                                                                                               
  namespace: openshift-etcd                                                                                                                                                                                     
spec:                                                                                                                                                                                                           
  ports:                                                                                                                                                                                                        
    - $patch: replace                                                                                                                                                                                           
    - name: etcd                                                                                                                                                                                                
      port: 2378                                                                                                                                                                                                
      protocol: TCP   

_**kustomization.yaml**_
apiVersion: kustomize.config.k8s.io/v1alpha1  # <-- Component notation                                                                                                                                          
kind: Component                                                                                                                                                                                                 
                                                                                                                                                                                                                
patches:                                                                                                                                                                                                        
- etcd-service-port.yaml                                                                                                                                                                                        
- etcd-host-service-port.yaml  

This can be rendered by kustomize. So we propose the installer to accept this kustomized output as input:

```kustomize build generatedManifests/ | openshift-install create cluster```

Reply:
@dhellmann do you want me to add this example to the enhancement too?

Reply:
@oglok I think having a detailed example in the body of the enhancement would be a good addition.

One of the technical challenges I see with implementing this is that, as far as I know, the installer treats manifests as files rather than individual resources. Since a file can contain multiple resources, and the names of the files are both significant and not necessarily derived from the contents in an obvious way, it will be challenging to turn a single input stream from stdin into the right artifacts to replace the cached versions on the filesystem. Which is not to say it can't be done, just that it isn't obvious to me how to do it. Have you started looking at the implementation of this idea already?

Reviewing the way that kustomize is used in the cluster-baremetal-operator repository, I see that it is configure to create output files in a named directory instead of streaming the results to stdout. Using that approach may make it easier to integrate with installer. A user would run `create manifests`, then `kustomize` with the manifests directory as input, then `create cluster`.

Reply:
Ok, I will add the example to the enhancement.

We haven't started looking at the implementation, but after one of the meetings we had with @crawford , he said it should be doable.

Reply:
In order for this to work, the installer would need to start keying assets based on their type and name rather than their on-disk path (since there aren't any paths when reading assets from STDIN). There's probably some annoying gotchas in this approach, but it doesn't fundamentally alter the design of the installer.